### PR TITLE
test and support for datatype in eval expressions

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?239879

### CODE REVIEW

**Comment by czue** [comment]
@dannyroberts buddy. probably easiest commit-wise

File: corehq/apps/userreports/expressions/specs.py [hunk 1]
```diff
@@ -258,14 +258,17 @@ class EvalExpressionSpec(JsonObject):
...
 
     def __call__(self, item, context=None):
         var_dict = self.get_variables(item, context)
         try:
-            return eval_statements(self.statement, var_dict)
+            untransformed_value = eval_statements(self.statement, var_dict)
+            transform = transform_from_datatype(self.datatype)
+            return transform(untransformed_value) if transform else untransformed_value
         except (InvalidExpression, SyntaxError, TypeError, ZeroDivisionError):
             return None
```
Comment (dannyroberts) [inline corehq/apps/userreports/expressions/specs.py]: What's the convention/contract here? If there's something wrong with the expression then return `None`?

File: corehq/apps/userreports/expressions/specs.py [hunk 2]
```diff
@@ -267,8 +267,7 @@ def __call__(self, item, context=None):
         var_dict = self.get_variables(item, context)
         try:
             untransformed_value = eval_statements(self.statement, var_dict)
-            transform = transform_from_datatype(self.datatype)
-            return transform(untransformed_value) if transform else untransformed_value
+            return transform_from_datatype(self.datatype)(untransformed_value)
```
Comment (dannyroberts) [inline corehq/apps/userreports/expressions/specs.py]: I guess `transform_from_datatype` never returns `None`?

**Review by dannyroberts — APPROVED** [review]
Looks good!

File: corehq/apps/userreports/expressions/specs.py (hunk 2 above)
Comment (czue) [inline corehq/apps/userreports/expressions/specs.py]: yeah it default to an identity function.

File: corehq/apps/userreports/expressions/specs.py (hunk 1 above)
Comment (czue) [inline corehq/apps/userreports/expressions/specs.py]: yep.

**Comment by dannyroberts** [comment]
great, just waiting on build then

**Pull request merged**